Batches page

rin-st opened the pull request:
Batches Page

- batches infinite table
- possibility to search by name
- creation of the batch
- editing the batch

Also, 
- moved `InfiniteTable` to separate component and used it for both batches and users pages
- used one component for both modals
- only admin can update/create batches
- if non-admin tries to see `/batches` page for now `null` returned. We need to figure out how to restrict only-admin client pages, I think for different PR
- if some required field is not filled showing notification
- updated tg field in schema to `notNull` + added migration


Creating batch:

<details><summary>Details</summary>
<p>

https://github.com/user-attachments/assets/282b3cf9-249a-439b-8ef1-30d8ff79d27d

</p>
</details> 



Updating batch + some sorting:

<details><summary>Details</summary>
<p>

https://github.com/user-attachments/assets/cd0319cd-44ce-4240-ba09-8324a2eed330

</p>
</details> 


To test: 
`yarn drizzle-kit studio` and then change your role to Admin.
Check `/batches` page

First PR of #125

carletex commented on packages/nextjs/services/database/migrations/0002_lovely_agent_brand.sql:
You ran the migration twice?

![image](https://github.com/user-attachments/assets/626b2aa0-c4b8-4e2d-9319-4578c7badf41)

`0002` was already taken.

Also, in case you haven't seen it in #128 https://github.com/drizzle-team/drizzle-orm/discussions/1104#discussioncomment-10814609

rin-st replied:
I did the same, but forgot [3]. Updated, thanks!

carletex commented on packages/nextjs/app/api/batches/[batchId]/update/route.ts:
```diff
@@ -0,0 +1,65 @@
+      return NextResponse.json({ error: "Missing required fields" }, { status: 400 });
+    }
+
+    const isAdmin = await isUserAdmin(address);
+    if (!isAdmin) {
+      return NextResponse.json({ error: "Only admins can update batches" }, { status: 403 });
+    }
```
I think this (super important) check is easy to miss in the current setup when we are doing routes.

Just wondering if we could improve it a bit:

Idea 1: Have `/api/admin/whatever` routes and enforce the check (middleware?)
Idea 2: Having a router wrapper?

```tsx
export const PUT = withAdminProtection(async (request, { params : {batchId: string} }) => {
    //
});
```

----

Just thinking out loud, not sure if it makes total sense or if NextJS allows it:

It seems that they don't recommend middleware for this?

https://nextjs.org/docs/app/building-your-application/routing/middleware

![image](https://github.com/user-attachments/assets/9218be35-174a-46d0-9244-f607549ceee2)


---

Anyway, just want to hear your thoughs, since this is an important issue (we can also do another PR for this)

carletex commented:
This is good stuff, Rinat. Thank you!

It turned a bit big (with the nice refactors) but I think it's good since this doesn't touch user-facing pages (like the portfolio page)

> if non-admin tries to see /batches page for now null returned. We need to figure out how to restrict only-admin client pages, I think for different PR

Yep, this is a main piece. Let's discuss in another issue, but let me throw a couple of options here:

1. Made "signature read": same as we do writes... but with reads (a bit painful if you refresh)
2. SIWE for admins

I'll let @technophile-04 do a first technical pass and @phipsae a functionality pass.

technophile-04 commented on packages/nextjs/services/database/repositories/batches.ts:
```diff
@@ -1,19 +1,105 @@
+
 export async function createBatch(batch: BatchInsert) {
-  const result = await db.insert(batches).values(batch).returning();
+  const countResult = await db.select({ count: sql<number>`count(*)` }).from(batches);
+  const currentCount = Number(countResult[0].count);
+
+  const result = await db
+    .insert(batches)
+    .values({
+      ...batch,
+      id: currentCount + 1,
+    })
+    .returning();
```
Umm we don't need to get the count? Since we have `id: serial().primaryKey()` in scheam it will autoincremented right?

rin-st replied:
For some reason if we don't have this, it doesn't work, when adding batch it says something like PrimaryKey (3) already exist. Looks like drizzle bug. Thought about using something like uuid but I use that id for generating website and github urls

carletex replied:
yes, leaving the id empty should automatically do it :/

> Thought about using something like uuid but I use that id for generating website and github urls

We shouldn’t rely on auto-increment IDs for this (like Batch #5), because they are not guaranteed to be sequential without gaps. If we import and then rollback, or if we add a batch and then remove it.... the database keeps an internal number of the autoincremental id *, so we could end up with numbers starting from a number != 0 and skipping. I just used a serial (instead of randon uuid) because we used serial in another table (but we should treat it as one. If it's confusing, we can change it)

* it's possible to restart that number, but in any case, we shouldn't use that ID

Even if we make sure it's always start from 0 and we don't skip any number.....something that would still happen: We create a "Stanford University Batch" (it gets assigned id = 16), and then we create "Batch 16" (id = 17)

So, options:
1. Add the github / website as fields (as you suggested to me last week)
2. We add an extra field (internalNumbering) for our typical internal batches.
3. We use the display name as the generator. e.g. "Batch 16" => lowercase / no space => batch16.buidlguidl.com

3 is a bit hacky, yes haha We could have a dynamic text below the admin form (Batch Name field) that shows the result live:

input: Bat => shows below: bat.buidlguidl.com
input: Batch 16 => shows below: batch16.buidlguidl.com
input: Standford Batch => shows below: standfordbatch.buidlguidl.com

But 1 & 2 make sense too.

phipsae replied:
I like the 3rd option :)

carletex replied:
Fix looks good. In https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137#discussion_r2053662414 I was suggesting allowing editing the field (bgWebsite), but we can do that in the future if needed.

Thanks!

rin-st replied:
dicided to add it anyway

bg subdomain field + check for validity

<details><summary>Details</summary>
<p>


https://github.com/user-attachments/assets/a58d4f16-b68a-4b19-b8d8-36b14c900c25


</p>
</details>

rin-st commented on packages/nextjs/services/database/repositories/batches.ts:
```diff
@@ -1,19 +1,105 @@
 
+export async function getBatchByName(name: string) {
+  return await db.query.batches.findFirst({
+    where: eq(batches.name, name),
```
Thanks, fixed [44f6778](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/44f677858d4b7a0692aa872c410e5816c1cc0243) + changed name search filter

phipsae commented:
Works great, thx @rin-st!

technophile-04 commented on packages/nextjs/services/api/batches/index.ts:
```diff
@@ -0,0 +1,96 @@
+  contractAddress,
+  telegramLink,
+  bgSubdomain,
+}: {
+  address: string;
+  signature: string;
+  name: string;
+  startDate: string;
+  status: string;
+  contractAddress?: string;
+  telegramLink: string;
+  bgSubdomain: string;
+}) {
```
Ohh I think maybe you should take this type as `CreateBatchPayload` from `/api/batches/create/route.ts` similar for editBatch?

rin-st replied:
thanks, updated [4300a41](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/4300a41a714d23457eadfa42ce411fedd1c5a77a)

technophile-04 commented on packages/nextjs/services/api/batches/index.ts:
```diff
@@ -0,0 +1,96 @@
+  return data.batch as Batch;
+}
+
+export async function editBatch(
```
Shall we call this `updateBatch` ?  similar to other functions like `updateLocation` in `services/api/users.ts`

rin-st replied:
I thought about it, but named `edit` for distinction between client modals components, so modalContent will not have the same name as editModal
<img width="1089" alt="image" src="https://github.com/user-attachments/assets/34e23085-c56a-4c5f-b6d1-4163c7b2db68" />

technophile-04 replied:
Umm lol yeah I see, I am like 60% on `updateBatch` in this file since like it a services or util irrespective of where it's called. So maybe naming it based on client code isn't that much worth it? Lol I mean it's being used at only place that too in client code so no strong opinion and we could keep it as it 🙌

technophile-04 commented on packages/nextjs/app/batches/_components/UpdateBatchModalContent.tsx:
```diff
@@ -0,0 +1,191 @@
+  }) => void;
+};
+
+export const UpdateBatchModalContent = forwardRef<HTMLInputElement, UpdateBatchModalContentProps>(
```
nit: Not sure update is the right word? Lol but also can't come with better name. 

Since this modal is used for both create and update batch

technophile-04 replied:
Ahh I see, we have this to display Conents like "UI" only component. 

Shall we call this just "BatchModalContent"? Because its UI only and updattion / creation functions are passed to it? 

And maybe rename EditModal to UpdateBatchModal? And also inline stuff with https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137#discussion_r2054303239?

technophile-04 commented on packages/nextjs/services/database/migrations/meta/0002_snapshot.json:
Ohh lol was wondering why this file was edited, it seems that you hit "save" in this file and you editor autoformatted this file, maybe you can copy it from `main` branch. So that we don't get diff of it. 

Lol that's what I did in last PR

rin-st replied:
thanks, updated [92c3874](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/92c3874d1b2c05e820e1cf77dfb115933217a7d4)

carletex commented on packages/nextjs/services/database/migrations/meta/0002_snapshot.json:
I think there are 2 snapshot files, there should only be 1, right?

rin-st replied:
it's just formatting for 0002

technophile-04 replied:
Ohh yeah just added a comment regarding it  https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137#discussion_r2054245559 🙌

rin-st replied:
thanks, updated [92c3874](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/92c3874d1b2c05e820e1cf77dfb115933217a7d4)

technophile-04 reviewed (commented):
Tysm @rin-st!! Just added some small comments and overall it looks great!! 

I tested locally and everything seems to work nicely! 

Maybe we could merge this PR after #136 and maybe incase we decide to keep autoincrement #143? 

Saying this because after than we could "branch" the database and also test it in staging environment with large data, especially the Infinite table (PS: love the abstaction of it!) 

But on local works nicely!! 

---

Some small nits on mobile device: 

<img width="554" alt="Screenshot 2025-04-22 at 8 03 50 PM" src="https://github.com/user-attachments/assets/3fe02a50-fe4c-46e4-b6b2-55fcee3b8b3c" />

And also builders page: 

<img width="203" alt="Screenshot 2025-04-22 at 8 04 20 PM" src="https://github.com/user-attachments/assets/f288ba54-ff82-4d65-84bd-cf660536e8be" />

Ohh lol just saw it's broken on main website too, and since this is useful for admin sides only we can update it in other PR 🙌

rin-st commented:
> Maybe we could merge this PR after https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/136 and maybe incase we decide to keep autoincrement https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/143?

Yes, and while waiting I'll fix debounce and will try to handle that table colors, thanks

rin-st commented:
> Maybe we could merge this PR after https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/136 and maybe incase we decide to keep autoincrement https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/143?

Backmerged main and returned autoincrement instead of `uuid` so it will be easier to merge after #143

rin-st commented on packages/nextjs/app/batches/page.tsx:
```diff
@@ -0,0 +1,217 @@
+        queryKey={tableQueryKey}
+        queryFn={getSortedBatches}
+        initialSorting={tableInitialSorting}
+        filter={filter}
```
[2771f5b](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/2771f5b165fe5da22e59786a6909ab2709ada70a)

rin-st commented:
- added filter debounce
- fixed table bg

<details><summary>Details</summary>
<p>

https://github.com/user-attachments/assets/7ee5d6cf-6ecd-460e-8fd9-76df432b8796

</p>
</details> 

and subdomain field https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137#discussion_r2055811848

carletex commented:
> This also reminds me that I need to tweak the import batch data script (to add the bgSubdomain field)

This is already done here. Nice, thanks!

carletex commented:
> we merge https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/148 (which makes sure all the users are present from BG).
> We create a PR and the following:
> first run 20250416_batch_data.ts script which makes sure all the batches had the subdomain and since prior we imported > all the users (https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/148) we won't face issue mentioned in
> We create a migration that remove .deafult("") from bgSubdomains

95 % Agree with this:
- I don't think we need the default "" at all
- I'll add a comment on #148 with a naming nitpicking :D

carletex commented on packages/nextjs/app/batches/page.tsx:
Can we move this to `/admin/batches`?

rin-st replied:
yep, moved [091bc04](https://github.com/BuidlGuidl/SpeedRunEthereum-v2/pull/137/commits/091bc04681d5d9141a88b458fc280e6abbf66630)

rin-st commented:
As I understand I don't need to do anything with default subdomain value and migrations, right?

technophile-04 commented:
> As I understand I don't need to do anything with default subdomain value and migrations, right?

Ohh no no, will just try it out and update the db with migrate and merge this, Thanks Rinat 🙌!!
